packages/server/src/server.js

```diff
@@ -111,21 +116,41 @@ export const invoke = async (invocation, server) => {
   const method = /** @type {string} */ (path.pop())
   const handler = resolve(server.service, path)
   if (handler == null || typeof handler[method] !== 'function') {
-    return /** @type {API.Result<any, API.HandlerNotFound>} */ (
-      new HandlerNotFound(capability)
-    )
+    return await Receipt.issue({
+      issuer: server.id,
+      ran: invocation,
+      result: {
+        /** @type {API.HandlerNotFound} */
+        error: new HandlerNotFound(capability),
+      },
+    })
   } else {
     try {
-      return await handler[method](invocation, server.context)
-    } catch (error) {
-      const err = new HandlerExecutionError(
+      const value = await handler[method](invocation, server.context)
+      return await Receipt.issue({
+        issuer: server.id,
+        ran: invocation,
+        // handler returns result in a different format from the receipt
+        // so we convert it here. We also need to handle the case where
+        // the handler `null` or `undefined` is returned which in receipt
+        // form at is unit type `{}`.
+        result: /** @type {API.ReceiptResult<{}>} */ (
+          value?.error ? { error: value } : { ok: value == null ? {} : value }
+        ),
+      })
+    } catch (cause) {
+      const error = new HandlerExecutionError(
         capability,
-        /** @type {Error} */ (error)
+        /** @type {Error} */ (cause)
```

Comment:
It feels brittle to type cast this here. There's no guarantee that `cause` matches `Error` afaict.